File: Heap (Priority Queue)/1792. Maximum Average Pass Ratio.cpp
#define pi pair<double, int>
class Solution {
public:
    double maxAverageRatio(vector<vector<int>>& classes, int extraStudents) 
    {
        // Lambda to compute delta (gain in ratio)
        auto delta = [&](int pass, int total) -> double {
            return (double(pass + 1) / (total + 1)) - (double(pass) / total);
        };
        
        int n = classes.size();
        priority_queue<pi> maxHeap;
        for (int i = 0; i < n; i++) 
        {
            int pass = classes[i][0], total = classes[i][1];
            maxHeap.push({delta(pass, total), i});
        }

        while (extraStudents--) 
        {
            auto [gain, idx] = maxHeap.top(); maxHeap.pop();
            int pass = classes[idx][0], total = classes[idx][1];
            
            classes[idx][0]++, classes[idx][1]++;
            maxHeap.push({delta(classes[idx][0], classes[idx][1]), idx});
        }

        double passRatio = 0.0;
        for (auto &cls : classes) 
            passRatio += double(cls[0]) / cls[1];

        return passRatio / n;
    }
};

File: 1792. Maximum Average Pass Ratio.py
class Solution:
    def delta(self, pass_count, total):
        return ((pass_count + 1) / (total + 1)) - (pass_count / total)
    def maxAverageRatio(self, classes: List[List[int]], extraStudents: int) -> float:
        maxHeap = []
        for passed, total in classes:
            heappush(maxHeap, (-self.delta(passed, total), passed, total))
        
        for _ in range(extraStudents):
            delta, passed, total = heappop(maxHeap)
            passed += 1
            total += 1
            heappush(maxHeap, (-self.delta(passed, total), passed, total))
        
        result = 0.0
        while maxHeap:
            _, passed, total = heappop(maxHeap)
            result += passed / total
        
        return result / len(classes)